synthtool/gcp/gapic_bazel.py

```diff
@@ -0,0 +1,236 @@
+from synthtool import metadata
+from synthtool import shell
+from synthtool.sources import git
+
+GOOGLEAPIS_URL: str = git.make_repo_clone_url("googleapis/googleapis")
+GOOGLEAPIS_PRIVATE_URL: str = git.make_repo_clone_url("googleapis/googleapis-private")
+LOCAL_GOOGLEAPIS: Optional[str] = os.environ.get("SYNTHTOOL_GOOGLEAPIS")
+
+
+class GAPICBazel:
+    """A synthtool component that can produce libraries using bazel build.
+    """
+
+    def __init__(self):
+        self._ensure_dependencies_installed()
+        self._googleapis = None
+        self._googleapis_private = None
+
+    def py_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "python", **kwargs)
+
+    def go_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "go", **kwargs)
+
+    def node_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "nodejs", **kwargs)
+
+    def csharp_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "csharp", **kwargs)
+
+    def php_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "php", **kwargs)
+
+    def java_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "java", **kwargs)
+
+    def ruby_library(self, service: str, version: str, **kwargs) -> Path:
+        return self._generate_code(service, version, "ruby", **kwargs)
+
+    def _generate_code(
+        self,
+        service: str,
+        version: str,
+        language: str,
+        *,
+        private: bool = False,
+        proto_path: Union[str, Path] = None,
+        output_dir: Union[str, Path] = None,
+        bazel_target: str = None,
+    ):
+        # Determine which googleapis repo to use
+        if not private:
+            googleapis = self._clone_googleapis()
+        else:
+            googleapis = self._clone_googleapis_private()
+
+        # Sanity check: We should have a googleapis repo; if we do not,
+        # something went wrong, and we should abort.
+        if googleapis is None:
+            raise RuntimeError(
+                f"Unable to generate {service}, the googleapis repository"
+                "is unavailable."
+            )
+
+        # Determine where the protos we are generating actually live.
+        # We can sometimes (but not always) determine this from the service
+        # and version; in other cases, the user must provide it outright.
+        if proto_path:
+            proto_path = Path(proto_path)
+            if proto_path.is_absolute():
+                proto_path = proto_path.relative_to("/")
+        else:
+            proto_path = Path("google/cloud") / service / version
+
+        # Determine bazel target based on per-language patterns
```

Comment:
This logic is complicated and single-purpose: please put it into its own function.